_posts/2018-02-21-symfony-route-test.md

```diff
@@ -0,0 +1,112 @@
+        // Add random/last static routes..
+
+        $output->writeln('First dynamic route: '.$this->matchRoutes(['/admin/statistics/3aa63971-1f61-4b28-bd60-5bb1c9582c8a/show']));
+        // Add random/last dynamic routes..
+        
+        $output->writeln('Not Found: '.$this->matchRoutes(['/foo/bar/baz']));
+    }
+
+    private function matchRoutes(array $routes): float
+    {
+        $startTime = microtime(true);
+        foreach ($routes as $route) {
+            for ($i = 0; $i < self::NR_OF_MATCHES; ++$i) {
+                $this->router->match($route);
+            }
+        }
+        $time = microtime(true) - $startTime;
+
+        return number_format(1000 * $time / count($routes), 2);
+    }
+{% endhighlight %}
+
+### Symfony 3.4
+In Symfony 3.4, the matcher has to iterate through all/most of the configured routes and try to find the route for the given url. 
+Trying to match a url towards the end of the list will result in an increased matching time, due to the number of comparisons 
+being done before the route is found. 
+
+That is why our tests for random route and last route take significantly more time to match than the first route. 
+The routes that were not found will take the most amount of time to match, since all the configured routes must be iterated
+before drawing the conclusion that our route does not exist. 
+
+### Symfony 4.1
+Instead of making separate `preg_match()` calls for each route, it combines all the regular expressions into a single regular expression.
+This means that we only have to call `preg_match()` once, and that is the biggest factor for faster matching. 
+
+## Results
+The table below shows how long it took to match the given routes 50.000 times. 
+
+The __Diff__ column displays how much faster the new Symfony router was, compared to our current one (3.4).  
+
+| Routes(ms)            | 3.4    | 4.1 (7d29a4d) | Diff  |
+| ----------------------|--------|---------------|-------|
+| First static route    | 448 ms | 382 ms        | -17%  |
+| Random static route   | 1621 ms| 474 ms        | -242% | 
+| Last static route     | 1826 ms| 544 ms        | -234% | 
+|                       |        |               |       |
+| First dynamic route   | 746 ms | 527 ms        | -41%  |
+| Random dynamic route  | 1454 ms| 531 ms        | -174% |
+| Last dynamic route    | 2039 ms| 524 ms        | -289% |
+|                       |        |               |       |
+| Not Found             | 2112 ms| 522 ms        | -304% | 
+
+As we can see, all of our routes were matched faster than previously. We especially see a gain in matching speed for 
+the random, last, and when the route was not found.  
+
+We are happy with these results, since the only thing we have to do once it is released is to 
+run `composer update`.    
+
+Remember that this is our results on our application. 
+
+Your results will probably not be the same since the percentage difference is highly dependant on the amount of routes
```

Comment:
dependent (with and "e")
on the amount of routes AND on the tree structure of the routes
my 77x boost is based on a pretty pathological tree layout for the 3.4 router
but since real route collections are quite balanced trees, the boost is lower